### PR TITLE
threads: add `shared` heap types

### PR DESCRIPTION
**Pull request by abrown**
This changes heap types everywhere to allow them to be `shared` as described in the shared-everything-threads [proposal]. The key refactoring is to push down all the abstract heap types (e.g., `func`, `any`, etc.) into an `AbstractHeapType`, leaving the `HeapType` enum with two variants: `Abstract` and `Concrete`. This allows the `Abstract` side to be marked `shared`; the `Concrete` side's `shared`-ness must be checked by looking at the pointed-to type, something which isn't yet addressed by this PR.

There are several parts of the PR that bump up against the boundary of what can be marked `shared` currently (e.g., `FuncType` and related types can't yet be marked `shared`). I've commented these locations with `TODO: handle shared` to be dealt with in future PRs that continue expanding `shared` everywhere.

[proposal]: https://github.com/WebAssembly/shared-everything-threads

### CODE REVIEW

File: crates/wasm-encoder/src/core/types.rs
Comment (abrown) [inline crates/wasm-encoder/src/core/types.rs:353]: If we think this is too wordy, I'm open to suggestions: I considered a helper like `HeapType::abstract(false, AbstractHeapType::func)` but decided to just directly construct this explicitly everywhere. But it's an option...

File: crates/wasm-mutate/src/mutators/peephole/dfg.rs
Comment (alexcrichton) [inline crates/wasm-mutate/src/mutators/peephole/dfg.rs:774]: I think these'll technically need to match `shared: false` as well.

Also if you're up for it adding `HeapType::FUNC` might be a reasonable constant to add to avoid expanding it like this.

File: crates/wit-component/src/gc.rs
Comment (alexcrichton) [inline crates/wit-component/src/gc.rs:1142]: I think this can be replaced with `let ty = ty.into()`

File: crates/wasmparser/src/validator/types.rs
Comment (alexcrichton) [reply crates/wasmparser/src/validator/types.rs:2769]: for example one way to organize this could be:

```
(A, List | Of | As | Supertypes) => true,
(A, _) => false,
// ...
```

so that way if a new `B` is added someone would be forced to look at this and consider if `A` was a supertype/subtype of `B`

File: crates/wasm-smith/src/core/code_builder.rs
Comment (alexcrichton) [reply crates/wasm-smith/src/core/code_builder.rs:6672]: (ditto for most things above too)

File: crates/wasm-encoder/src/core/types.rs
Comment (alexcrichton) [reply crates/wasm-encoder/src/core/types.rs:353]: I think this is fine, it's just "the one place" to write this down for the `pub const` values

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:1561]: This reminds me that I've been meaning for some time now to clean up all the decoding here in this crate since all the various constants all over the place feel a bit overkill.

**Review by alexcrichton — COMMENTED** [review]
High-level question, how complete do you want to get this before landing? For example do you want to have all the `// TODO` resolved?

File: crates/wast/src/core/binary.rs
Comment (abrown) [reply crates/wast/src/core/binary.rs:317]: They shouldn't have, they're just grouped under a single pattern at new lines 329-332. I.e., if we have a ref that is nullable and a heap type, then print this byte encoding (using the `Encode for AbstractHeapType` in this case). I did this same kind of thing in several places. The one change in behavior from before is that, in this short-form case, we prepend `0x65` if the abstract heap type is shared.

File: crates/wasmparser/src/validator/operators.rs
Comment (abrown) [reply crates/wasmparser/src/validator/operators.rs:3997]: Opened https://github.com/WebAssembly/shared-everything-threads/issues/65 to track this. Not sure how I can create a `HeapType::ANY` from `enum HeapType`? Some `const` form?

File: crates/wasmprinter/src/lib.rs
Comment (abrown) [reply crates/wasmprinter/src/lib.rs:1027]: Yeah, that's what it would appear that we are settling on for the long-form of shared abstract references. I.e., `(shared funcref)` == `(ref null (shared func))`.

Comment (alexcrichton) [reply crates/wasmprinter/src/lib.rs:1027]: Ok sounds good 👍

File: crates/wast/src/core/binary.rs
Comment (alexcrichton) [reply crates/wast/src/core/binary.rs:317]: Ah ok I think I got myself confused, carry on!

File: crates/wasmparser/src/validator/operators.rs
Comment (alexcrichton) [reply crates/wasmparser/src/validator/operators.rs]: I think you should be able to do `impl HeapType { pub const ANY: HeapType = /* ... */; }` to get the `::ANY` working

File: crates/wasm-smith/src/core/code_builder.rs
Comment (abrown) [reply crates/wasm-smith/src/core/code_builder.rs:5389]: Yeah, I figured this helper function (here and elsewhere) is a way to start down that path but clearly more will be needed once it comes time to actually turn everything on in wasm-smith.

File: crates/wasmparser/src/readers/core/types.rs
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs]: I think you were looking at an old, un-rebased version. The latest code has this fixed.

File: crates/wasmparser/src/readers/core/types.rs
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:1598]: Looks like that happened in #1601; resolving this.

File: crates/wasmparser/src/readers/core/types.rs
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:1187]: Sure. I have tried this a couple of times and it's not as easy as it might seem. How about I remove the various levels of this and switch the logic over to a `WasmModuleResources` function in another PR?

File: crates/wasmparser/src/readers/core/types.rs
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:924]: I mean, what I'm trying to do here is factor out the common pattern, mainly for human comprehensibility. I would just use `RefType::wat` but (a) it's behind the `validate` feature and (b) because of its `&'static str` return it can't print out the concrete index like we were accustomed to here. If we dropped (a) and made `RefType::wat` return a `String` then we can just have one version of all this...

Perhaps better yet, why don't we just use `Debug` everywhere `RefType::wat` was being used (i.e., `ty_to_str`)?

File: crates/wast/src/core/types.rs
Comment (abrown) [reply crates/wast/src/core/types.rs:84]: I haven't added `*.wast` tests for this PR because of the issue with checking shared-ness of concrete types. Until I can sort out how to properly check if a type is shared, I can't validate any of this. I did mention below that I wanted to try that in a separate PR, but I guess we should decide to either:
- disable `*.wast` parsing and accept this PR without tests
- fix the shared check and include the `*.wast` tests in this PR.

**Comment by abrown** [comment]
> High-level question, how complete do you want to get this before landing? For example do you want to have all the // TODO resolved?

No, I think I want to keep the boundary-level `TODO: handle shared` around so I can make sure to check all these places when I expand shared types even more. The real question for me about completeness is testing: as I describe above, I can't write spec tests without shared-checking, but I kind of wanted to defer that for a future PR. I'm open to fixing it all here, though, but it's all getting a bit unwieldy...

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [reply crates/wasmparser/src/readers/core/types.rs:924]: Keeping the `&str` is somewhat important for performance in the operand validator because that's used in a few places for error messages and not dealing with `Debug` or `String` helps codegen/perf there.

This may just be me not really being familiar enough with the gc proposal. I basically just keep forgetting all the various shorthands like `nullexnref` works but `(ref nullexn)` doesn't or something like that. Anyway just ignore me here I think.

**Review by alexcrichton — COMMENTED** [review]
My main worry with this is that there's no tests currently for this PR and personally at least once I land a `// TODO: ...` it's highly likely that I just forget about it and don't actually handle it until years later when someone hits a bug. I do think that some tests should be part of this, at the very least parsing/printing/dumping/etc or something along those lines. The tests don't necesarily need to be exhaustive, but I think there should be at least some simple things working at this point, even if they should be invalid modules in the future or something like that.

As for things like `is_shared`, if you're ok taking on the responsibility to audit the `// TODO: ...` after this lands I think that's ok.

File: crates/wasmparser/src/readers/core/types.rs
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs]: Done, see ad37d62.

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**
